catalog/app/utils/checksums.ts

```diff
@@ -0,0 +1,91 @@
+import * as B64 from 'js-base64'
+import pLimit from 'p-limit'
+
+import cfg from 'constants/config'
+import * as Model from 'model'
+
+// 8 MiB -- boto3 default: https://boto3.amazonaws.com/v1/documentation/api/latest/reference/customizations/s3.html#boto3.s3.transfer.TransferConfig
+const MIN_PART_SIZE = 1024 ** 2 * 8
+
+const MAX_PARTS = 10000 // Maximum number of parts per upload supported by S3
+
+function getPartSize(fileSize: number): number | null {
+  // use single-part upload (and plain SHA256 hash)
+  if (!cfg.useMultipartChecksums || fileSize < MIN_PART_SIZE) return null
+
+  // NOTE: in the case where fileSize is exactly equal to MIN_PART_SIZE
+  // boto creates a 1-part multipart upload :shrug:
+  let partSize = MIN_PART_SIZE
+  let numParts = Math.ceil(fileSize / partSize)
+
+  while (numParts > MAX_PARTS) {
+    partSize *= 2
+    numParts = Math.ceil(fileSize / partSize)
+  }
+
+  return partSize
+}
+
+const toHex = (buf: ArrayBuffer) =>
+  Array.from(new Uint8Array(buf))
+    .map((b) => b.toString(16).padStart(2, '0'))
+    .join('')
+
+const singlepart = (value: ArrayBuffer): Model.PackageEntryHash => ({
+  value: toHex(value),
+  type: Model.HASH_TYPE_SP,
+})
+
+const multipart = (value: ArrayBuffer, partCount: number): Model.PackageEntryHash => ({
+  value: `${B64.fromUint8Array(new Uint8Array(value))}-${partCount}`,
+  type: Model.HASH_TYPE_MP,
+})
+
+function mergeBuffers(buffers: ArrayBuffer[]) {
+  const buf = new Uint8Array(buffers.reduce((acc, cur) => acc + cur.byteLength, 0))
+  let offset = 0
+  for (const arr of buffers) {
+    buf.set(new Uint8Array(arr), offset)
+    offset += arr.byteLength
+  }
+  return buf
+}
+
+// limit concurrency: 2 files / 8 blobs at a time
+// XXX: benchmark and fine-tune this
+const fileLimit = pLimit(2)
+const blobLimit = pLimit(8)
+
+async function hashBlob(blob: Blob) {
+  const buf = await blob.arrayBuffer()
+  return window.crypto.subtle.digest('SHA-256', buf)
+}
+
+const hashBlobLimit = (blob: Blob) => blobLimit(hashBlob, blob)
+
+async function computeFileChecksum(f: File): Promise<Model.PackageEntryHash> {
```

Comment:
I'd pass `window.crypto.subtle` as an argument to simplify unit-test, because node environment has a slightly different crypto module

Reply:
I mean, `computeFileChecksum(file, window.crypto.subtle)` in browser and `computeFileChecksum(file, crypto.webCrypto.subtle)` in test environment

Reply:
But it just me, I don't like to mock `window` or anything else for tests :)

Reply:
i'd just mock that if/when necessary